Topic/thread reschedule

jamshark70 opened the pull request:
## Purpose and Motivation

New feature suggested in #4903: Make it possible to reschedule a Routine, Task or EventStreamPlayer transparently onto a different clock or to a different (later) time.

Along the way, I found out that our documentation for `nextTimeOnGrid` is quite poor. I haven't updated it yet. It's tangentially related to this PR, so I'm not sure if I should just fold it in here or make another PR. (`nextTimeOnGrid` doesn't change substantially here, except for a new `referenceBeat` argument. Default behavior is unchanged.)

(There is one unrelated documentation cleanup: Task help said that Condition could not be used within a Task. That hasn't been true for some years, so I took the liberty of removing that remark.)

## Types of changes

- Documentation
- New feature

(The `nextTimeOnGrid` third argument could be a breaking change if someone has created an extension with additional arguments for this method selector. AFAICS within the main class library, there is no breaking change.)

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

telephon commented:
A really good new feature, thank you. I'll test it.

telephon commented on SCClassLibrary/Common/Core/Thread.sc:
```diff
@@ -110,6 +111,27 @@ Routine : Thread {
+			};
+			clock = argClock;
+		};
+		if(this === thisThread) {
```
if there is a good name for what happens here, I think this should be a method on function, similar to `forkIfNeeded`

jamshark70 replied:
How would it look? If it's a method of Function, then you wouldn't have implicit access to the thread as `this`, so maybe it would have to be like this, in context of this PR: `{ ... }.deferFromThread(this/*.threadPlayer*/)`. (I'm not sure this is the place to bikeshed a new method though.)

telephon replied:
>I'm not sure this is the place to bikeshed a new method though

As a non native speaker, the meaning of bikeshedding remains obscure (I'm pretty sure I've never done it, although I do have a bicycle ;-)). 


Joke aside, this definitely is the place to discuss how to factor a problem. This may include anew method, and the general usefulness of such a method. If it turns out to be a large discussion, we can still do this in a separate thread.

I think your second version is good, if we assume that the intention is to _defer away from_ or _defer off_  a given thread. This method makes clear what reading the code leaves obscure.

jamshark70 replied:
OK, I've done that (and even instinctively named the method `deferAwayFrom`).

I have to admit, writing the documentation for the new method renewed my doubts about it (though its usages in both `reschedule` implementations do look cleaner). But, let's leave that to the review process.

jamshark70 commented:
> Joke aside, this definitely is the place to discuss how to factor a problem.

It's a delicate balance. We've already seen some PRs spiral out of control. This PR isn't in danger of that... I think I probably err on the side of keeping PRs close to the initial scope.

I don't mind to change it here. It'll be a couple of days later.

telephon commented:
> It's a delicate balance. We've already seen some PRs spiral out of control.

This forum is not necessarily the adequate place for developer cooperation, as you said, it depends.

jamshark70 commented:
I'd like to add the WIP tag for now -- I just realized that this is implemented for streams based on Routines, but not for any variety of FuncStream. That would be an exceptional usage but for completeness, it should be there.

jamshark70 commented:
> I just realized that this is implemented for streams based on Routines, but not for any variety of FuncStream. That would be an exceptional usage but for completeness, it should be there.

I'm afraid I will have to give up on this. I couldn't figure out how to get the intrinsic variables to work.

Currently, we have `Routine : Thread : Stream : AbstractFunction : Object` (cf. `Routine.superclasses` = same). Object, AbstractFunction and Stream have no instance variables (slots). `Thread` has a lot of them, and they are defined as intrinsic variables in PyrObject.cpp. (`Routine` doesn't declare any additional ones.)

This PR adds `rescheduledTime` to Thread (inherited by Routine) -- so, `reschedule` is defined for Thread, and separately for `PauseStream`.

FuncStream inherits from Stream, not Thread. That suggests moving `var rescheduledTime` to Stream -- which means moving one of the intrinsic variables. I couldn't figure out how to do this without crashing sclang on startup. (For that matter, it isn't clear how intrinsic variables work with inheritance.)

I'm also reconsidering that maybe this isn't a big deal. You can play a Pfunc, and rescheduling works because it's really an EventStreamPlayer that's accessing the FuncStream.

```supercollider
(
p = Pfunc {
	var delta = exprand(0.05, 0.25);
	(midinote: rrand(48, 72), dur: delta, legato: 0.1)
}.play;
)

p.reschedule(quant: 3);  // works

p.stop;
```

But you can't do the same with FuncStream:

```supercollider
(
p = FuncStream({
	var delta = exprand(0.05, 0.25);
	(midinote: rrand(48, 72), dur: delta, legato: 0.1).play;
	delta  // return/reschedule
}).play;
)

p.reschedule(quant: 3);  // Message 'reschedule' not understood

thisProcess.stop;  // it's not a stream player so, no 'stop'
```

... but in almost 20 years, I've never seen anyone do this -- not least because there's no convenient way to stop it.

I had been thinking of the case of `p = aPattern.collect { ... }.play` but that's a Pcollect, analogous to `Pfunc` above which is already working -- and `p = aPattern.asStream.collect { ... }.play` doesn't work at all (and never did) -- so, you always need the pattern wrapper anyway.

Except here:

```
r = Routine {
	var delta;
	loop {
		delta = exprand(0.05, 0.25);
		(midinote: rrand(48, 72), dur: delta, legato: 0.1).play;
		delta.wait
	}
};

q = r.collect { |x| x * 2 }.play;

q.stop;  // doesn't stop!
thisProcess.stop;  // so have to do this :-\
```

... but nobody does that either, because of the same problem (no way to stop `q`).

TL;DR except for abstracting out `deferFromThread` (in a moment...), I think we can leave this as this, and move forward to review.

telephon commented:
I agree that it would be too much to expect the full scheduling abilities on `FuncStream`. In general, it is always a good idea to improve uniformity in the stream system, but this feature is in line with other, similar particularities, as you say. So let's go ahead with this as it is.

telephon commented on HelpSource/Classes/Task.schelp:
```diff
@@ -49,6 +44,13 @@ Stop playing now. (Pause and stop have the same implementation.)
+
+Switch the Task to a different clock, or a different time, without stopping. See link::Classes/Routine#-reschedule:: for complete documentation.
+
+NOTE:: If you want to reschedule a Task from within the Task itself, code::thisThread.reschedule(...):: will not work (as it does in a Routine). You must write code::thisThread.threadPlayer.reschedule(...):: instead.
```
Maybe better: 
If you want to reschedule a Task from within the Task itself, code::thisThread.reschedule(...):: will not work, because the task is what plays the thread. This is unlike in a Routine, which plays itself.

telephon commented on SCClassLibrary/Common/Core/Clock.sc:
```diff
@@ -298,11 +298,12 @@ elapsed time is whatever the system clock says it is right now. elapsed time is
-		if (quant < 0) { quant = beatsPerBar * quant.neg };
-		if (phase < 0) { phase = phase % quant };
-		^roundUp(this.beats - baseBarBeat - (phase % quant), quant) + baseBarBeat + phase
+	nextTimeOnGrid { arg quant = 1, phase = 0, referenceBeat;
```
shouldn't the new argument of `nextTimeOnGrid` be added to all other implementations of this method? E.g. there is `IdentityDictionary:nextTimeOnGrid`.

jamshark70 commented on HelpSource/Classes/Routine.schelp:
```diff
@@ -217,6 +217,87 @@ fork {
+list::
+## Rescheduling to a different clock will be continuous in terms of seconds, but not necessarily continuous in terms of beats. Time-based patterns such as link::Classes/Pseg::, or the use of link::Classes/Env:: as a stream, are likely to break.
+
+When switching the clock, code::reschedule:: will first calculate the new "awake" time, in beats, based on the given code::quant::. Then it converts that code::beats:: value into the new clock's code::beats:: for the same instant in time. If the old clock's tempo is 1 and the Routine was waiting for 1 beat, the next "awake" should happen after one second. When rescheduling without a code::quant:: ...
```
First question, then, is: Do you think I should explain it in complete, exhaustive detail, or just give a high-level overview and let users infer the behavior from examples?

The complete explanation goes something like this:

There are four relevant moments in time:

A. When the thread *last* awoke.
B. Now.
C. When the thread is *next* scheduled to awake.
D. The rescheduled time. This may be the same moment as C, but translated to a different clock; or later than C (on the same or a different clock).

While it's supported to reschedule a thread from within the thread, often the `reschedule` caller will be outside the thread. In that case, B is not constrained to line up with any of the routine times: A <= B < C <= D (if B == C, behavior might be undefined).

> If it "was waiting", then the time has passed -- it should happen instantly?

"... was waiting for 1 beat" means: at A, a `1.0.wait` call occurred. So we know C = A + 1, and B must be within 1 beat of A.

Rescheduling does not change C -- even if you reschedule onto a different clock, the thread will awake at time C. I think that's why I used past tense here -- you can't change a previously issued `wait` (whereas, if I said it "is waiting," you might think you could make it "be waiting" for a different length of time). (And, if D == translateToOtherClock(C), then the thread will continue from the same moment when it would have awakened, but with a new `beats` base and possibly a different tempo.)

I think it should be clear by now why I'm struggling with the right balance between clarity and concision. The above is too much detail for one method's documentation -- but trying to streamline the discussion ends up confusing some crucial details.

jamshark70 commented:
I believe I addressed the review comments. (And, on second thought, I changed that "was waiting" to "is waiting" -- maybe it's OK now? Though that section is difficult to explain.)

telephon commented:
Thank you, this reads very well, and the `nextTimeOnGrid` solution for `IdentityDictionary` is also convincing.

jamshark70 commented:
> the nextTimeOnGrid solution for IdentityDictionary is also convincing.

I'd like to take another look at this interface tomorrow, to be sure it will really work for rescheduling.

jamshark70 commented:
I think my suspicion that last night was correct -- my change to `IdentityDictionary:nextTimeOnGrid` was not right.

TempoClock:nextTimeOnGrid calls `quant.asQuant` -- so, whatever object is returned by IdentityDictionary:asQuant needs to follow the same method signature as Quant:nextTimeOnGrid, namely `aQuant.nextTimeOnGrid(clock, referenceBeat)`.

IdentityDictionary says `asQuant { ^this.copy }`. Therefore it should be `anIdentityDictionary.nextTimeOnGrid(clock, referenceBeat)` -- the reference beat must be passed in at the time of resolving the quant. (Also, the reference beat is not a property of a Quant object -- so it shouldn't be a property of a dictionary being used as a quant. *Editorial* Confusion introduced by adding a new feature without a proper functional spec -- I know we'll never really start writing functional specs before doing code changes, but it really does concretely cause problems :rofl: )

Committed a fix.

telephon commented:
In my experience, the best thing is always a page of concise examples.

joshpar commented:
@telephon @jamshark70 - sorry about the delay on this PR - what is the current consensus? Should we pull this in to 3.13, and if so, anything else needed?

jamshark70 commented:
> what is the current consensus? Should we pull this in to 3.13, and if so, anything else needed?

At this point, I don't remember. It generally looks OK...?

telephon commented:
Yes, I think it is good.

dyfer commented:
I resolved a merge conflict in the helpfile (I think).

I assume this is okay to merge, but because the conversation was stalled a while back I'll wait a bit before pulling this @telephon @jamshark70

jamshark70 commented:
IIRC this is ready...? Thanks for fixing the conflict.

woutersnoei commented:
Hi folks, there seems to be an issue with this merge, considering PauseStream/Task when playing on AppClock. Reproduce like this:
Task({ "hello".postln }, AppClock ).play;

--> ERROR: Message 'schedAbs' not understood.

This is because AppClock doesn't understand schedAbs..

dyfer commented:
@jamshark70 @telephon what do you think about the issue brought up by @woutersnoei ?

telephon commented:
hm, I never really understood why `AppClock` doesn't understand `schedAbs`. I've run into this before …

dyfer commented:
> there seems to be an issue with this merge, considering

@woutersnoei after seeing response from @telephon I checked and this issue was introduced in the current development version of SC, _after_ 3.12, but _before_ this PR. It is reported in #5700. It would be good to fix it before releasing 3.13 as it is a regression (AFAIU). Let's continue the conversation in #5700.
